pkg/gather/collect/podcollector.go

```diff
@@ -0,0 +1,323 @@
+		},
+	}
+
+	return pc
+}
+
+func retrieveContainerLogs(ctx context.Context, podClient corev1client.PodInterface, destinationPath string, podName string, logOptions *corev1.PodLogOptions) error {
+	dest, err := os.OpenFile(destinationPath, os.O_CREATE|os.O_WRONLY, 0666)
+	if err != nil {
+		return fmt.Errorf("can't open file %q: %w", destinationPath, err)
+	}
+	defer func() {
+		err := dest.Close()
+		if err != nil {
+			klog.ErrorS(err, "can't close file", "Path", destinationPath)
+		}
+	}()
+
+	logsReq := podClient.GetLogs(podName, logOptions)
+	readCloser, err := logsReq.Stream(ctx)
+	if err != nil {
+		return fmt.Errorf("can't create a log stream: %w", err)
+	}
+	defer func() {
+		err := readCloser.Close()
+		if err != nil {
+			klog.ErrorS(err, "can't close log stream", "Path", destinationPath, "Pod", podName, "Container", logOptions.Container)
+		}
+	}()
+
+	_, err = io.Copy(dest, readCloser)
+	if err != nil {
+		return fmt.Errorf("can't read logs: %w", err)
+	}
+
+	return nil
+}
+
+func (c *PodCollector) collectContainerLogs(ctx context.Context, logsDir string, podMeta *metav1.ObjectMeta, podCSs []corev1.ContainerStatus, containerName string) error {
+	var err error
+
+	cs, _, found := oslices.Find(podCSs, func(s corev1.ContainerStatus) bool {
+		return s.Name == containerName
+	})
+	if !found {
+		klog.InfoS("Container doesn't yet have a status", "Pod", naming.ObjRef(podMeta), "Container", containerName)
+		return nil
+	}
+
+	var limitBytes *int64
+	if c.logsLimitBytes > 0 {
+		limitBytes = pointer.Ptr(c.logsLimitBytes)
+	}
+
+	logOptions := &corev1.PodLogOptions{
+		Container:  containerName,
+		Timestamps: true,
+		Follow:     false,
+		LimitBytes: limitBytes,
+	}
+
+	// TODO: Tolerate errors in case state changes in the meantime (like when a pod is being restarted in backoff)
+	//       It's error prone to just ignore it, maybe we should retry and refreshing the state and retrying instead.
+	//       https://github.com/scylladb/scylla-operator/issues/1400
+
+	if cs.State.Running != nil {
+		// Retrieve current logs.
+		logOptions.Previous = false
+		err = retrieveContainerLogs(ctx, c.corev1Client.Pods(podMeta.Namespace), filepath.Join(logsDir, containerName+".current"), podMeta.Name, logOptions)
+		if err != nil {
+			return fmt.Errorf("can't retrieve pod logs for container %q in pod %q: %w", containerName, naming.ObjRef(podMeta), err)
+		}
+	}
+
+	if cs.LastTerminationState.Terminated != nil {
+		logOptions.Previous = true
+		err = retrieveContainerLogs(ctx, c.corev1Client.Pods(podMeta.Namespace), filepath.Join(logsDir, containerName+".previous"), podMeta.Name, logOptions)
+		if err != nil {
+			return fmt.Errorf("can't retrieve previous pod logs for container %q in pod %q: %w", containerName, naming.ObjRef(podMeta), err)
+		}
+	}
+
+	return nil
+}
+
+func (c *PodCollector) Collect(ctx context.Context, u *unstructured.Unstructured, resourceInfo *ResourceInfo) error {
```

Comment:
How about reordering the functions in the file so that the public ones come right after the constructor? That way, someone reading it for the first time can "read the code like prose", starting from the top-level functions and jumping to their callees as needed.